### PR TITLE
feat: add new decorators filter-in and filter-out

### PR DESCRIPTION
**Pull request by IgorKarpiuk**
## What/Why/How?
 Add new decorator filter-in and filter-out
## Reference

Closes https://github.com/Redocly/redocly-cli/issues/599

## Check yourself

- [x] Code is linted
- [ ] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

**Comment by IgorKarpiuk** [comment]
@RomanHotsiy Is it the expected result?

## Filter-in

#### Input yaml

```yaml
     openapi: 3.0.0
     paths:
        /pet:
            x-audience: [Public, Global]
            post:
              summary: test
              requestBody:
                content:
                  application/x-www-form-urlencoded:
                    schema:
                      type: object
          /user:
            x-audience: [Public, Protected]
            post:
              summary: test
              requestBody:
                content:
                  application/json:
                    schema:
                      type: object  
```
#### Decorator config

```yaml
    lint:
      decorators:
        filter-in:
          property: x-audience
          value: [Public, Global]
          matchStrategy: all
```

#### Result

```yaml
     openapi: 3.0.0
     paths:
        /pet:
            x-audience: [Public, Global]
            post:
              summary: test
              requestBody:
                content:
                  application/x-www-form-urlencoded:
                    schema:
                      type: object
```

## Filter-out

#### Input yaml

```yaml
     openapi: 3.0.0
     paths:
        /pet:
            x-audience: [Private, Protected]
            post:
              summary: test
              requestBody:
                content:
                  application/x-www-form-urlencoded:
                    schema:
                      type: object
          /user:
            x-audience: [Public, Global]
            post:
              summary: test
              requestBody:
                content:
                  application/json:
                    schema:
                      type: object  
```
#### Decorator config

```yaml
    lint:
      decorators:
        filter-out:
          property: x-audience
          value: [Private, Protected]
          matchStrategy: all
```

#### Result

```yaml
     openapi: 3.0.0
     paths:
         /user:
           x-audience: [Public, Global]
           post:
             summary: test
             requestBody:
               content:
                 application/json:
                   schema:
                     type: object  
```

File: packages/core/src/decorators/common/filters/filter-helper.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    if (strategy === 'all') {
+        return decoratorValue.every( (item: string) => nodeValue.includes(item));
+    }
+    return false;
+}
+
+function toLowerCaseArrayIfNeeded(value: string | string[]): string[] {
+    if (!Array.isArray(value)) {
+        value = [value.toLowerCase()];
```
Comment (RomanHotsiy) [inline packages/core/src/decorators/common/filters/filter-helper.ts:69]: why do we lowercase? let's do case-sensitive comparison

File: packages/core/src/decorators/__tests__/filter-in.test.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+          get:
+            parameters:
+              - $ref: '#/components/parameters/x'
+        /user:
+          access: private
+          get:
+            parameters:
+              - $ref: '#/components/parameters/x'      
+      components:
+        parameters:
```
Comment (RomanHotsiy) [inline packages/core/src/decorators/__tests__/filter-in.test.ts:48]: Would be great to add another param `y` here to test if it gets removed.

File: packages/core/src/decorators/__tests__/filter-in.test.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    const testDoc = parseYamlToDocument(
+      outdent`
+        swagger: '2.0'
+        host: api.instagram.com
+        paths:
+          '/geographies/{geo-id}/media/recent':
+            get:
+              parameters:
+                - description: The geography ID.
+                  access: private
```
Comment (RomanHotsiy) [inline packages/core/src/decorators/__tests__/filter-in.test.ts:208]: fix to `x-access` plz

File: packages/core/src/decorators/common/filters/filter-helper.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    return false;
+}
+
+function toArrayIfNeeded(value: string | string[]): string[] {
+    if (!Array.isArray(value)) {
+        value = [value];
+    }
+    return value;
+}
```
Comment (RomanHotsiy) [inline packages/core/src/decorators/common/filters/filter-helper.ts:72]:
```suggestion
function toArrayIfNeeded<T>(value: T | T[]): T[] {
   return Array.isArray(value) ? value : [value];
}
```

**Review by RomanHotsiy — COMMENTED** [review]
Looks pretty good.

Would be great to add one more test where filter-in properties are on different levels, e.g. on pathItem and on parameters.

**Review by RomanHotsiy — APPROVED** [review]
LGTM 🚢

File: packages/core/src/decorators/common/filters/filter-helper.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
+import {UserContext} from "../../../walk";
+import {isRef} from "../../../ref-utils";
+import {isEmptyArray, isEmptyObject, isPlainObject} from "../../../utils";
```
Comment (SmoliyY) [inline packages/core/src/decorators/common/filters/filter-helper.ts:3]: Please use prettier for this and rest new files.

**Pull request merged**